src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -3001,16 +3003,30 @@ HCIMPL3(VOID, JIT_Unbox_Nullable, void * destPtr, CORINFO_CLASS_HANDLE type, Obj
 HCIMPLEND
 
 /*************************************************************/
-/* framed helper that handles full-blown type equivalence */
-NOINLINE HCIMPL2(LPVOID, JIT_Unbox_Helper_Framed, CORINFO_CLASS_HANDLE type, Object* obj)
+/* framed Unbox helper that handles enums and full-blown type equivalence */
+NOINLINE HCIMPL2(LPVOID, Unbox_Helper, CORINFO_CLASS_HANDLE type, Object* obj)
 {
     FCALL_CONTRACT;
 
     LPVOID result = NULL;
 
+    TypeHandle typeHnd(type);
+    // boxable types have method tables
+    _ASSERTE(!typeHnd.IsTypeDesc());
+    MethodTable* pMT1 = typeHnd.AsMethodTable();
+    MethodTable* pMT2 = obj->GetMethodTable();
+
     OBJECTREF objRef = ObjectToOBJECTREF(obj);
     HELPER_METHOD_FRAME_BEGIN_RET_1(objRef);
-    if (TypeHandle(type).IsEquivalentTo(objRef->GetTypeHandle()))
+
+    if (pMT1->GetInternalCorElementType() == pMT2->GetInternalCorElementType() &&
```

Comment:
Moving this inside `HELPER_METHOD_FRAME` will regress performance for enums...

Reply:
We return a ref to the boxed value. If I want to poll, I need to protect the box. Is there a way to protect without a frame?

---
In reply to: [379777249](https://github.com/dotnet/runtime/pull/32353#discussion_r379777249) [](ancestors = 379777249)

Reply:
One way to do this is to use fast path without HMF when `g_TrapReturningThreads.LoadWithoutBarrier()` is zero, and use polling show path with HMF otherwise. It would duplicate a bit of the logic, but it should not be a huge problem for the unbox helper.

Reply:
Another way to do this would be to introduce `FC_GC_POLL_EX` variant for byrefs, but that would be quite a bit of plumbing. I do not think it is worth it.

Reply:
Right. A flag check would not trash the refs and should be much cheaper.

Reply:
With suggested changes (`g_TrapReturningThreads` and `GetVerifierCorElementType`) the enum case is slightly faster than originally. 

That is - as long as `g_TrapReturningThreads` is checked _after_ comparing  `GetVerifierCorElementType`.  
Not sure why the order of compares matters. It is nearly the same codegen. I am guessing the rarely taken branch is more predictable this way.